jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java

```diff
@@ -516,61 +644,139 @@ private boolean updateForwardedHandle(MethodHandles.Lookup lookup, String header
         return !_handles.put(headerName, lookup.findVirtual(Forwarded.class, forwardedMethodName, type));
     }
 
-    private static class ForcedHostPort extends HostPort
+    private static class MutableHostPort
     {
-        ForcedHostPort(String authority)
+        public static final int UNSET = -1;
+        public static final int IMPLIED = 0;
+
+        String _host;
+        Priority _hostPriority = Priority.UNSET;
+        int _port = UNSET;
+        Priority _portPriority = Priority.UNSET;
+
+        public void setHost(String host, Priority priority)
         {
-            super(authority);
+            if (priority.ordinal() > _hostPriority.ordinal())
+            {
+                _host = host;
+                _hostPriority = priority;
+            }
         }
-    }
 
-    private static class PossiblyPartialHostPort extends HostPort
-    {
-        PossiblyPartialHostPort(String authority)
+        public void setPort(int port, Priority priority)
         {
-            super(authority);
+            if (priority.ordinal() > _portPriority.ordinal())
+            {
+                _port = port;
+                _portPriority = priority;
+            }
         }
 
-        protected PossiblyPartialHostPort(String host, int port)
+        public void setHostPort(HostPort hostPort, Priority priority)
         {
-            super(host, port);
+            if (priority.ordinal() > _hostPriority.ordinal())
+            {
+                _host = hostPort.getHost();
+                _hostPriority = priority;
+            }
+
+            int port = hostPort.getPort();
+            // Is port supplied?
+            if (port > 0 && priority.ordinal() > _portPriority.ordinal())
+            {
+                _port = hostPort.getPort();
+                _portPriority = priority;
+            }
+            // Since we are Host:Port pair, the port could be unspecified
+            // Meaning it's implied.
+            // Make sure that we switch the tracked port from unset to implied
+            else if (_port == UNSET)
+            {
+                // set port to implied (with no priority)
+                _port = IMPLIED;
+            }
         }
-    }
 
-    private static class PortSetHostPort extends PossiblyPartialHostPort
-    {
-        PortSetHostPort(String host, int port)
+        @Override
+        public String toString()
         {
-            super(host, port);
+            final StringBuilder sb = new StringBuilder("MutableHostPort{");
+            sb.append("host='").append(_host).append("'/").append(_hostPriority);
+            sb.append(", port=").append(_port);
+            sb.append("/").append(_portPriority);
+            sb.append('}');
+            return sb.toString();
         }
     }
 
-    private static class Rfc7239HostPort extends HostPort
+    /**
+     * Ordered Priority Enum.
+     * <p>
+     * Lowest priority first.
+     * </p>
+     */
+    public enum Priority
```

Comment:
Rather than "Priority" I think "Source" is a better name for this.  Will make more sense in logging to see something like "source=XFORWARDED_FOR".  It also matches a lot of our webapp config stuff where we record the source, which is then used to determine the priority.